Sources/CacheAdvance/CacheReader.swift

```diff
@@ -44,29 +41,29 @@ final class CacheReader {
         reader.offsetInFile
     }
 
-    /// Returns the next encodable message, seeking to the beginning of the next message.
-    func nextEncodedMessage() throws -> Data? {
-        let startingOffset = offsetInFile
-
-        guard startingOffset != offsetInFileAtEndOfNewestMessage else {
-            // We're at the last message.
-            return nil
+    /// Returns the encodable messages in a range
+    ///
+    /// - Parameter startOffset: the start offset to start reading
+    /// - Parameter endOffset: the end offset to stop reading, the default value is EOF if the value is nil.
+    func encodedMessagesFromOffset(_ startOffset: UInt64, endOffset: UInt64? = nil) throws -> [Data] {
+        var encodedMessages = [Data]()
+        try reader.seek(to: startOffset)
+        while let data = try nextEncodedMessage() {
+            encodedMessages.append(data)
+            if let endOffset = endOffset, offsetInFile >= endOffset {
```

Comment:
I'm thinking about the case where `offsetInFile == endOffset` to make sure I agree that we want to break in that case.

When we write a message we update the offset in file at the end of the newest message to the offset of the writer after the write.

https://github.com/dfed/CacheAdvance/blob/54e61e242d51e7239bdcf9f4307cd8355b7e86ba/Sources/CacheAdvance/CacheAdvance.swift#L247-L252

I believe that the offset at the end of the newest message is the same as the offset of the start of the next range.

This seems to be confirmed by the [documentation](https://developer.apple.com/documentation/foundation/filehandle/1410936-write) of the write method.

<img width="1056" alt="Screen Shot 2022-11-07 at 2 49 04 PM" src="https://user-images.githubusercontent.com/1791049/200431924-e08ce6c0-c6a2-4cff-8563-b1321358a202.png">

Reply:
@dfed is this the correct documentation to reference? I was surprised to see we are using a deprecated method, though I recall that there was some reason we were doing so.

Reply:
these are the `write` methods we're using:
https://github.com/dfed/CacheAdvance/blob/3a1ed6f99c2a870dc09c3a8144a58f915757a569/Sources/CacheAdvance/FileHandleExtensions.swift#L33-L40

I should take a pass to update the `__` methods in this file and use the more modern ones. We are not using deprecated methods when newer ones are available.

Reply:
Took a pass at updating the methods we use here: https://github.com/dfed/CacheAdvance/pull/74